### PR TITLE
Remove hydration hocs in favour of apiFetch Middlewares

### PR DESCRIPTION
**Pull request by mikejolley**
Gutenberg provides an apiFetch middleware called `createPreloadingMiddleware` which accepts REST API data. This is very very similar to the hydration code we implemented, but more straightforward to use. 

This PR passes endpoint data to `createPreloadingMiddleware` and removes our custom HOC implementation. This also removes the need to handle stale cart data on page load (implemented in https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3960) because `createPreloadingMiddleware` does not exhibit the same problem.

**Testing**

1. Go to the cart block page. Check there were no extra `fetch` requests to `wc/store/cart` in the network inspector.
2. Go to the checkout block page. Ensure there is no error and checkout works as normal.

Testing stale data:

- Go to cart block
- Change cart item quantities. Take note of the amounts.
- Browse to a different page
- Use the browser back button.
- Confirm amounts are still correct and not stale.

@opr confirmed this removes the need for the fix in #3960 
@Aljullu could also use your eyes since it touched a few parts of the mini-cart block

### CODE REVIEW

File: src/Assets/AssetDataRegistry.php
Comment (mikejolley) [inline src/Assets/AssetDataRegistry.php:354]: createPreloadingMiddleware does the job our hocs were doing previously.

File: src/BlockTypes/Checkout.php
Comment (mikejolley) [inline src/BlockTypes/Checkout.php:342]: Checkout still needs manual preloading data because it doesn't go via wp.data. It should, but doesn't yet.

File: src/StoreApi/Schemas/CartSchema.php
Comment (mikejolley) [inline src/StoreApi/Schemas/CartSchema.php:380]: This should have been a header, like the product collection `Last-Modified` header handling. But I found it wasn't needed anyway...

**Comment by mikejolley** [comment]
@senadir you might want a look also

**Comment by opr** [comment]
Hi @mikejolley looks like the changes implemented in #3960 _aren't_ needed anymore, I tested it and it works, the newest cart is always fetched when navigating back and forward using the browser buttons 🙌🏼 

I like this approach as it simplifies our codebase and is easier to reason with.

File: assets/js/base/context/providers/cart-checkout/checkout-state/constants.ts
Comment (mikejolley) [inline assets/js/base/context/providers/cart-checkout/checkout-state/constants.ts:37]: Checkout requires preloading still - it uses getSetting and does not have a  wp data store like the cart does.

File: assets/js/blocks/cart-checkout/mini-cart/with-mini-cart-conditional-hydration.tsx
Comment (senadir) [inline assets/js/blocks/cart-checkout/mini-cart/with-mini-cart-conditional-hydration.tsx:1]: I'd prefer if @Aljullu take a look at this.

File: tests/e2e/specs/frontend/cart.test.js
Comment (senadir) [inline tests/e2e/specs/frontend/cart.test.js:1]: Why was this deleted?

**Review by senadir — APPROVED** [review]
Nothing like the smell of freshly deleted code. However, I think you deleted a bit too much with `tests/e2e/specs/frontend/cart.test.js`?

I'm pre-approving on the basis that this tests fine, and the code looks good, but I will wait for a review from @Aljullu to confirm the minicart related part.

File: tests/e2e/specs/frontend/cart.test.js
Comment (opr) [reply tests/e2e/specs/frontend/cart.test.js:1]: This test file was only testing the code relating to reloading a "stale" cart - since this PR removes this code, and the possibility for carts to go stale the tests are redundant.

Comment (senadir) [reply tests/e2e/specs/frontend/cart.test.js:1]: Lovely!

File: assets/js/blocks/cart-checkout/mini-cart/with-mini-cart-conditional-hydration.tsx
Comment (Aljullu) [reply assets/js/blocks/cart-checkout/mini-cart/with-mini-cart-conditional-hydration.tsx:1]: Mini Cart block changes LGTM. :ok_hand:

**Comment by mikejolley** [comment]
@senadir the tests I removed were for the localstorage timestamp that was removed.

**Pull request merged**
